File: src/bomipay/schemas/provider.py
from typing import Dict, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, EmailStr, constr, field_serializer


class ProviderCredentials(BaseModel):
    api_key: constr(min_length=1)
    secret_key: constr(min_length=1)


class ProviderConnectRequest(BaseModel):
    merchant_id: Optional[str]
    provider_name: constr(min_length=2, max_length=128)
    credentials: ProviderCredentials


class ProviderTestRequest(BaseModel):
    provider_name: constr(min_length=2, max_length=128)
    public_key: constr(min_length=1)
    secret_key: constr(min_length=1)
    webhook_secret: Optional[str] = None


class ProviderTestResponse(BaseModel):
    success: bool
    message: Optional[str] = None


class ProviderAccountData(BaseModel):
    provider_account_id: UUID | str
    provider_name: str
    status: str

    @field_serializer('provider_account_id')
    def serialize_uuid(self, value: UUID | str, _info) -> str:
        return str(value) if value else None


class ProviderConnectResponse(BaseModel):
    success: bool
    data: ProviderAccountData


class ProviderHealthResponse(BaseModel):
    provider_name: str
    merchant_id: UUID | str
    status: str
    connected: bool

    @field_serializer('merchant_id')
    def serialize_uuid(self, value: UUID | str, _info) -> str:
        return str(value) if value else None


class ProviderListResponse(BaseModel):
    provider_account_id: UUID | str
    provider_name: str
    merchant_id: UUID | str
    status: str

    model_config = ConfigDict(from_attributes=True)

    @field_serializer('provider_account_id', 'merchant_id')
    def serialize_uuid(self, value: UUID | str, _info) -> str:
        return str(value) if value else None

File: bomipay-website/src/lib/provider.ts
import api from './api'

export interface ProviderTestRequest {
  provider_name: string
  public_key: string
  secret_key: string
  webhook_secret?: string
}

export interface ProviderTestResponse {
  success: boolean
  message?: string
}

export interface ProviderConnectRequest {
  provider_name: string
  public_key: string
  secret_key: string
  webhook_secret?: string
  environment: 'test' | 'live'
}

export interface ProviderConnectResponse {
  success: boolean
  data: {
    provider_account_id: string
    provider_name: string
    status: string
  }
}

export async function testConnection(req: ProviderTestRequest): Promise<ProviderTestResponse> {
  const response = await api.post('/providers/test-connection', {
    provider_name: req.provider_name,
    public_key: req.public_key,
    secret_key: req.secret_key,
    webhook_secret: req.webhook_secret,
  })
  return response.data
}

export async function connectProvider(req: ProviderConnectRequest): Promise<ProviderConnectResponse> {
  const response = await api.post('/providers/connect', {
    provider_name: req.provider_name,
    credentials: {
      api_key: req.public_key,
      secret_key: req.secret_key,
    },
  })
  return response.data
}

export async function listProviders() {
  const response = await api.get('/providers')
  return response.data
}

export async function disconnectProvider(providerAccountId: string) {
  await api.delete(`/providers/${providerAccountId}`)
}

export async function getProviderHealth(providerName: string) {
  const response = await api.get(`/providers/${providerName}/health`)
  return response.data
}
